kernel/kernel-api/src/main/java/io/delta/kernel/internal/replay/ReverseFilesToColumnarBatchActionsIterable.java

```diff
@@ -0,0 +1,247 @@
+                    actionsIter.get().close();
+                    actionsIter = Optional.empty();
+                    closed = true;
+                }
+            }
+
+            /**
+             * If the current `actionsIter` has no more elements, this function finds the next
+             * non-empty file in `filesIter` and uses it to set `actionsIter`.
+             */
+            private void tryEnsureNextActionsIterIsReady() {
+                if (actionsIter.isPresent()) {
+                    // This iterator already has a next element, so we can exit early;
+                    if (actionsIter.get().hasNext()) {
+                        return;
+                    }
+
+                    // Clean up resources
+                    try {
+                        actionsIter.get().close();
+                    } catch (IOException ex) {
+                        throw new RuntimeException(ex);
+                    }
+
+                    // Set this to empty since we don't know if there's a next file yet
+                    actionsIter = Optional.empty();
+                }
+
+                // Search for the next non-empty file and use that iter
+                while (filesIter.hasNext()) {
+                    actionsIter = Optional.of(getNextActionsIter());
+
+                    if (actionsIter.get().hasNext()) {
+                        // It is ready, we are done
+                        return;
+                    }
+
+                    // It was an empty file. Clean up resources
+                    try {
+                        actionsIter.get().close();
+                    } catch (IOException ex) {
+                        throw new RuntimeException(ex);
+                    }
+
+                    // Set this to empty since we don't know if there's a next file yet
+                    actionsIter = Optional.empty();
+                }
+            }
+
+            /**
+             * Get the next file from `filesIter` (.json or .checkpoint.parquet), contextualize it
+             * (allow the connector to split it), and then read it + inject the `isFromCheckpoint`
+             * information.
+             *
+             * Requires that `filesIter.hasNext` is true.
+             */
+            private CloseableIterator<Tuple2<FileDataReadResult, Boolean>> getNextActionsIter() {
+                final FileStatus nextFile = filesIter.next();
+                final Path nextFilePath = new Path(nextFile.getPath());
+
+                try {
+                    if (nextFilePath.getName().endsWith(".json")) {
+                        final JsonHandler jsonHandler = tableClient.getJsonHandler();
+
+                        // Convert the `nextFile` FileStatus into an internal ScanFile Row and then
+                        // allow the connector to contextualize it (i.e. perform any splitting)
+                        final CloseableIterator<FileReadContext> fileReadContextIter =
```

Comment:
These closeable iterators should be added to a context which can be used to close them in case this function doesn't return successfully. For example create `fileReadContextIter` is successful, but not the fileReadDataIter create.

Reply:
> These closeable iterators should be added to a context 

What exactly does this mean? Put them in a try-with-resources? Expose them to the user?